eclipse-platform/eclipse.platform.ui
# Transfer primary data stack when current one is closed #1774

The "primaryDataStack" is the main editor part stack used for placing newly opened editors. When the shared area containing the `primaryDataStack` is split into multiple stacks and the last editor in the `primaryDataStack`is closed, the `primaryDataStack`is removed from the application model so that the application model does not contain a `primaryDataStack`anymore.

The cleanup functionality that removes a part stack being not the last in the shared are is placed in the `CleanupAddon`. This change adds functionality to this addon that transfers the `primaryDataStack`property from the stack being removed to the first of the remaining part stacks. It also centralizes the functionality to create a primary data stack and to identify an element as such to avoid duplicated functionality.

Note that this chance does _not_ yet solve the problem of the onboarding information being lost within the same session of your workbench. It will only ensure that the `primaryDataStack` does not get lost, such that when restarting the product the onboarding information will be back. The issue of disappearing onboarding information within the same session has to be addressed by changing the way in which the onboarding information is instantiated. I will provide a separate PR for that.

Contributes to https://github.com/eclipse-platform/eclipse.platform.ui/issues/1774

### How to test / reproduce

You can experience the change with the following scenario (also documented in #1774):
1. Start an Eclipse product, e.g., an SDK
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/755472/e1a71d08-c6f3-45ad-911e-71cf44d11672)
2. Open two editors in the shared area. Then split the shared area placing the the two editors side by side via drag and drop.
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/755472/40b92290-b9b6-4dbf-9cfd-b33bd04da87a)
3. Close the _left_ editor first, then close the remaining (the originally _right_) editor. You will see that the onboarding information is lost.
![image](https://github.com/eclipse-platform/eclipse.platform.ui/assets/755472/20568116-8fa7-4e1e-8a24-667a698aed32)
4. Restart the product. Without this chance, you will see that the onboarding information is still lost (like in the previous screenshot) whereas with this change the onboarding information is present again. This is a result of the `primaryDataStack` still being present. You can also investigate the `workbench.xmi` under `.metadata/.plugins7org.eclipse.e4.workbench` within your workspace. With this change, the `workbench.xmi` will still an entry like the following after executing the described scenario whereas without this chance the entry is missing:
```xml
    <sharedElements xsi:type="advanced:Area" xmi:id="_SYOZIOq3Ee6Ds-nHDao7XQ" elementId="org.eclipse.ui.editorss">
      <children xsi:type="basic:PartStack" xmi:id="_VvWtAeq3Ee6SwsvRjze8_w" elementId="org.eclipse.e4.primaryDataStack">
        <tags>org.eclipse.e4.primaryDataStack</tags>
        <tags>EditorStack</tags>
      </children>
    </sharedElements>
```
